src/odemis/odemisd/main.py

```diff
@@ -328,71 +329,98 @@ def _terminate_all_alive(self):
+        executor = futures.ThreadPoolExecutor(max_workers=20)
+        fs = set()
+        while self._parents:
+            parent_less = tuple(c for c, p in self._parents.items() if not p and not c in terminating_comps)
+            if not parent_less and not fs:
+                # just pick a random component
                 logging.warning("All the components to terminate have parents: %s",
-                                parents)
-                parent_less = (tuple(parents.keys())[0],)
-
-            for c in parent_less:
-                cname = c.name
-                logging.debug("Stopping comp %s", cname)
-                # TODO: update the .alive VA every time a component is stopped?
-                # maybe not necessary as we are finishing _everything_
-                try:
-                    # FIXME: don't try to terminate the children created by
-                    # delegation after terminating their parent (and container)
-                    # For now it just causes a warning message, but it'd be
-                    # better to not have the warning at all.
-                    c.terminate()
-                except Exception:
-                    logging.warning("Failed to terminate component '%s'", cname, exc_info=True)
-
-                # Terminate the container if that was the component for which it
-                # was created.
-                # TODO: check there is really no component still running in the
-                # container?
-                if cname in self._instantiator.sub_containers:
-                    container = self._instantiator.sub_containers[cname]
-                    logging.debug("Stopping container %s", container)
-                    try:
-                        container.terminate()
-                    except Exception:
-                        logging.warning("Failed to terminate container %r", container, exc_info=True)
-                    del self._instantiator.sub_containers[cname]
-
-                # remove from the graph
-                del parents[c]
-                for p in parents.values():
-                    p.discard(c)
-                alive.discard(c)
-                try:
-                    mic.alive.value = alive
-                except Exception:
-                    logging.warning("Failed to update the alive VA", exc_info=True)
+                                self._parents)
+                parent_less = (tuple(self._parents.keys())[0],)
+
+            for comp in parent_less:
+                terminating_comps.append(comp)
+                f = executor.submit(self._terminate_parentless, comp)
+                fs.add(f)
+
+            # Block until one future is completed, then continue looping
+            done, not_done = futures.wait(fs, return_when=futures.FIRST_COMPLETED)
+            fs = not_done
+
+        futures.wait(fs, return_when=futures.ALL_COMPLETED)
```

Comment:
Let's add a log here, saying that it's done requesting the termination of all components, and just still waiting for N to be done.